### ISSUE TITLE
Android 6 Compatibility

### ISSUE THREAD

**Issue by app2one** [mmazzarolo/react-native-beacons-android#15]
The module currently doesn't work on android 6.  It seems start from android 6 additional permission is required on runtime bluetooth usage.  Is anyone come up with any idea?

**Comment by mmazzarolo** [comment]
Hi! 
Thanks for the  info! I'm on holyday right now, I'll take a look at it in 7 days :)

**Comment by alessandro-bottamedi** [comment]
have you tried with this package?
https://github.com/lucasferreira/react-native-android-permissions

**Comment by mmazzarolo** [comment]
@alessandro-bottamedi Thanks, did you already worked with that module? Is it stable? 
[Also, other info on Android M permissions on RN](https://github.com/lwansbrough/react-native-camera/issues/224)

**Comment by app2one** [comment]
Yes. That module works. Thanks. 

從我的 iPhone 傳送

**Comment by alessandro-bottamedi** [comment]
@mmazzarolo 
It seems stable, Next week I will release an app in beta testing with this Android Permissions module,  I'll let you know...

**Comment by alessandro-bottamedi** [comment]
Here's what I think I understand:

In android 6.0 there are two different levels of permission: NORMAL and DANGEROUS.
For the NORMAL permission is sufficient to place them in the android manifest, while the DANGEROUS must have the permissions at runtime.

to detect beacons you must have bluetooth and localization active and accepted this permissions (I'm right??):

android.permission.BLUETOOTH and android.permission. BLUETOOTH_ADMIN are in NORMAL level so it need to be only in the android.manifest.
ACCESS_FINE_LOCATION and ACCESS_COARSE_LOCATION are DANGEROUS levels so it need the new android 6.0 runtime permission.

Here the list of permissions with its levels: https://developer.android.com/reference/android/Manifest.permission.html

To make everything work i've used this: [react-native-permissions](https://github.com/yonahforst/react-native-permissions) i've discarded react-native-android-permissions because this not have iOs support.

And i ask only for location permission in runtime it is sufficent for detecting ibeacon.

I'm working now to make the right procedure to check if the app have permission on android and iOs, and check the geolocation and bluetooth state...

Please correct me if i'm wrong

**Comment by mmazzarolo** [comment]
Thank you Alessandro (sei italiano anche tu, giusto?), 
it seems you're correct, if you're interested you can also check the [altbeacons SDK for android documentation about the Marshmellow permissions](http://altbeacon.github.io/android-beacon-library/samples.html).
From your understanding do you thing it would be possibile to handle the runtime permission from `react-native-beacons-android` itself instead of forcing the user to use another external dependency? 
The cleanest solution in my opinion should be adding the feature in react-native but it doesn't seem to be in its roadmap.  

And thank you again for your time man 👍

**Comment by alessandro-bottamedi** [comment]
(Si italiano, lo hai capito dal mio pessimo inglese o dal nome? :wink: )

The Estimote-Android-SDK docs has much more clear explanation:

> In order to reliably detect beacons following conditions must be met:
> - Bluetooth permissions are granted (android.permission.BLUETOOTH and android.permission.BLUETOOTH_ADMIN). This is done automatically if you use Estimote SDK. BeaconService is declared in AndroidManifest.xml. This is done automatically if you use Estimate SDK.
> - If running on Android M or later, Location Services must be turned on.
> - If running on Android M or later and your app is targeting SDK < 23 (M), any location permission (ACCESS_COARSE_LOCATION or ACCESS_FINE_LOCATION) must be granted for background beacon detection.
> - If running on Android M or later and your app is targeting SDK >= 23 (M), any location permission (ACCESS_COARSE_LOCATION or ACCESS_FINE_LOCATION must be granted.

Maybe you can insert an explanation like this in the README and the link for these Permissions Packages...

Anyway i think that the Permissions Package must be separated from the others, so you can use it more times in your app for all your permissions needs. But its only my opinion.
Preferably it should be insert in react-native of course!

**Comment by mmazzarolo** [comment]
> (Si italiano, lo hai capito dal mio pessimo inglese o dal nome? 😉 )

Dal nome, sono l'ultimo che può criticare l'inglese di qualcuno visto il mio, tranquillo!  

> Maybe you can insert an explanation like this in the README and the link for these Permissions Packages...

Yup, gimme two sec

**Comment by mmazzarolo** [comment]
Done! Closing this issue while waiting for an official solution, thanks @app2one and @alessandro-bottamedi for pointing out this issue.

Feel free to post here if needed 👍

**Comment by micheletedeschi** [comment]
FYI now react native (0.33.0) support android M permission!

**Comment by mmazzarolo** [comment]
@micheletedeschi Unexpected, nice!
Did you already try it? Any hint on the implementation of it on `react-native-beacons-android`? Thanks!

**Comment by micheletedeschi** [comment]
Yes and works! Only be sure to ask permission before start ranging or monitoring.

You can find a permission example here: [https://github.com/facebook/react-native/blob/0fb2ccfcc3e6babe28d7d1bcf4635ef2f9191eb1/Examples/UIExplorer/js/PermissionsExampleAndroid.android.js](url)

**Comment by maggialejandro** [comment]
Hi!
I'm trying to make this work on a Nexus 5, Android 6.0.1

>     "react": "15.3.2",
>     "react-native": "0.34.1",
>     "react-native-beacons-android": "^2.1.0",
>     "react-native-permissions": "^0.2.5"

npm 3.9.3
node 6.2.1

Also, I added these permissions to the manifest

>     <uses-permission android:name="android.permission.INTERNET" />
>     <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW"/>
>     <uses-permission android:name="android.permission.BLUETOOTH"/>
>     <uses-permission android:name="android.permission.BLUETOOTH_ADMIN"/>
>     <uses-permission android:name="android.permission.BLUETOOTH_PRIVILEGED"/>
>     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>

Here is the code
http://pastebin.com/HTqqZrg3

Any ideas ?
Thanks in advance
